apps/lily-wallet/docker-compose.yml

```diff
@@ -0,0 +1,34 @@
+version: '3'
+
+services:
+  frontend:
+    image: kaybesee/lily-frontend:latest
+    restart: on-failure
+    stdin_open: true
+    build:
+      dockerfile: Dockerfile.frontend
+      context: ./
+    ports:
+      - '42069:80'
+
+  express:
+    privileged: true
```

Comment:
Ideally the usb communication would be handled via browser APIs, but I haven’t been able to get the HWI library to work in a browser setting.

Your comment made me go down a long webusb rabbit hole just now though, and I think there might be a way to wrap HWI in some sort of wrapper to have it work in the browser. 

I’ve gotta do more digging on how that webusb spec works but I was able to detect and give browser access to my coldcard and Trezor tonight. Still fuzzy on how to send/receive messages and what messages to send to get xpub and signing data though. 

I get what you’re saying about being in another room and having to plug your HWW into the pi. It’s not an ideal experience, but I do think it’s good enough for an initial release. I imagine most people are running their servers somewhere in their home anyway and would only use their HWW sparingly after initial setup. 

Would having an official Umbrel HWI app be appropriate to maintain app sandboxing? That way Lily could hit some url and access device data that way. Other apps could do the same (I’m thinking Specter which currently require you to download an additional piece of bridge software to use with Umbrel). 

I do think the HWW is a crucial part of the app, so I’m fine holding off on releasing it until we can sandbox HWI into its own app or figure out wrapping HWI in a webusb wrapper.

Reply:
Hmmnn yeah, I agree. We can ship with direct USB access. However we still need to figure out how to do this without `privileged: true` or `volumes: /dev:/dev` since that basically means the Lily container has the same privilages as the host.

I think it might work with something like:

```
devices:
  - /dev/ttyUSB0
```

(not sure of this exactly, maybe some other similar path)

Which should allow us to keep the container unprivileged but just pass the USB device directly into the container, and _I think_ HWI should be able to then enumerate connected HWWs from inside the container.

However I think this might only work if the HWW is plugged in before the app starts and break if it's remove and re-attached while thew app is running.

I'm also not sure how portable this is, I'm not sure if that path is reliable across Linux distributions, but that's not a deal breaker as long as it works on Umbrel OS.

Some interesting reading here: https://stackoverflow.com/a/53892718

Would also be interested to hear @nevets963 thoughts on this.

Reply:
Some other good reading:

- http://marc.merlins.org/perso/linux/post_2018-12-20_Accessing-USB-Devices-In-Docker-_ttyUSB0_-dev-bus-usb-_-for-fastboot_-adb_-without-using-privileged.html
- https://docs.docker.com/engine/reference/commandline/create/#dealing-with-dynamically-created-devices---device-cgroup-rule

Looks like maybe this may not be trivial to get working reliably in a safe way. Unless the assumption is that the user leaves the HWW attached to their Umbrel.

Reply:
I see that there are HWI binaries to interface with USB for each platform within the image. Could you try running a container with the following args. and see if you're still able to enumerate/interact with USB devices using the appropriate HWI binary

```
docker run --device-cgroup-rule='c 189:* rmw' -v /dev/bus:/dev/bus:ro -it <image> bash
```

It's also important to test that everything still functions even after a device has been ejected and re-inserted.